### PR TITLE
Added Discount Codes to apib docs

### LINKED ISSUE #3732: Implement API Blueprint for documenting the nextgen API
Follow up to discussion on #3641

We will be using Api-Blueprint (https://apiblueprint.org/). This issue is a high priority one. Needs to be done by the end of this week.

### PR DESCRIPTION
**Pull request by enigmaeth**
#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes 

#### Short description of what this resolves:

Fixes #3732.

### CODE REVIEW

File: docs/api/api_blueprint.apib [hunk 1]
```diff
@@ -4423,6 +4422,327 @@ Delete a single notification.
...
+          "jsonapi": {
+              "version": "1.0"
+          },
+          "links": {
+              "self": "/v1/discount-codes/1"
+          }
+        }
+
+### Delete Notification [DELETE]
+Delete a single notification.
```
Comment (magdalenesuo) [inline docs/api/api_blueprint.apib]: @enigmaeth Delete Discount Code and Delete a single discount code XD

File: docs/api/api_blueprint.apib [hunk 2]
```diff
@@ -4422,6 +4422,327 @@ Delete a single notification.
           }
         }
 
+# Group Discount Codes
+Discount Codes related to the events.
+
+| Parameter | Description | Type | Required |
+|:----------|-------------|------|----------|
+| `code`  | Discount Code | string | - |
+| `discount_url` | Message/content of the notification | string | - |
```
Comment (SaptakS) [inline docs/api/api_blueprint.apib]: The parameters should all be dasherized.

**Comment by enigmaeth** [comment]
@magdalenesuo @SaptakS Updated.

File: docs/api/api_blueprint.apib [hunk 3]
```diff
@@ -4423,6 +4422,324 @@ Delete a single notification.
...
+|:----------|-------------|------|----------|
+| `code`  | Discount Code | string | - |
+| `discount-url` | URL of the discount | string | - |
+| `value` | Value of the discount | float | - |
+| `type` | Amount or Percentage | string | - |
+| `is-active` | If the discount code is active | boolean | - |
+| `tickets-number` | Tickets Number for Discount Code | integer (>=0) | - |
+| `min-quantity` | Minimum number of tickets for Discount Code | integer (>=0) | - |
+| `max-quantity` | Maximum number of tickets for Discount Code | integer (>=0) | - |
+| `type` | Amount or Percentage | string | - |
```
Comment (magdalenesuo) [inline docs/api/api_blueprint.apib]: @enigmaeth There are two `type` descriptions, remove one

File: docs/api/api_blueprint.apib [hunk 4]
```diff
@@ -4423,6 +4422,324 @@ Delete a single notification.
...
+| `used-for` | Discount Code used for | String | - |
+
+
+## Discount Code Collection [/events/{event_id}/discount-codes{?page%5bsize%5d,page%5bnumber%5d,sort,filter}]
++ Parameters
+    + event_id: 1 (integer) - ID of the specific user in integer
+    + page%5bsize%5d (optional, integer, `10`) - Maximum number of resources in a single paginated response.
+    + page%5bnumber%5d (optional, integer, `2`) - Page number to fetched for the paginated response.
+    + sort (optional, string, `name`) - Sort the resources according to the given attribute in ascending order. Append '-' to sort in descending order.
+    + filter (optional, string, `[{"is_read": "true" }]`) - Filter according to the flask-rest-jsonapi filtering system. Please refer: http://flask-rest-jsonapi.readthedocs.io/en/latest/filtering.html for more.
```
Comment (magdalenesuo) [inline docs/api/api_blueprint.apib]: @enigmaeth I do not think there is an is_read field in discount-code so this example is not proper. maybe make it `"type": "percentage"` or something similar

Comment (enigmaeth) [inline docs/api/api_blueprint.apib]: Resolved.

File: docs/api/api_blueprint.apib [hunk 5]
```diff
@@ -4423,6 +4422,323 @@ Delete a single notification.
...
+| `discount-url` | URL of the discount | string | - |
+| `value` | Value of the discount | float | - |
+| `type` | Amount or Percentage | string | - |
+| `is-active` | If the discount code is active | boolean | - |
+| `tickets-number` | Tickets Number for Discount Code | integer (>=0) | - |
+| `min-quantity` | Minimum number of tickets for Discount Code | integer (>=0) | - |
+| `max-quantity` | Maximum number of tickets for Discount Code | integer (>=0) | - |
+| `valid-from` | Discount Code valid from | ISO 8601 (tz-aware) | - |
+| `used-for` | Discount Code used for | String | - |
+
```
Comment (magdalenesuo) [inline docs/api/api_blueprint.apib]: @enigmaeth description for `tickets`, `created-at` and `valid-till` fields are missing

Comment (enigmaeth) [inline docs/api/api_blueprint.apib]: Resolved.

File: docs/api/api_blueprint.apib [hunk 6]
```diff
@@ -4422,6 +4421,327 @@ Delete a single notification.
...
+                  "discount-url": "https://my-discount-url.com",
+                  "value": "100",
+                  "type": "amount",
+                  "is_active": "false",
+                  "tickets_number": "404",
+                  "min_quantity": "0",
+                  "max_quantity": "100",
+                  "valid_from": "2017-06-18T18:30:00+00:00",
+                  "valid_till": "2017-06-24T18:30:00+00:00",
+                  "used_for": "Event"
```
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @enigmaeth 
These need to be dasherized.
min-quantity , max-quantity , valid-from , valid-till , used-for

File: docs/api/api_blueprint.apib [hunk 7]
```diff
@@ -4422,6 +4421,327 @@ Delete a single notification.
...
+                  },
+                  "attributes": {
+                      "tickets": null,
+                      "code": "DC101",
+                      "valid-from": "2017-06-18T18:30:00+00:00",
+                      "min-quantity": 0,
+                      "created-at": "2017-06-20T09:59:23.740772+00:00",
+                      "tickets-number": 404,
+                      "value": 100,
+                      "max-quantity": 100,
```
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @enigmaeth 
Shouldn't the `tickets-number` be <=100 ?

File: docs/api/api_blueprint.apib [hunk 8]
```diff
@@ -4422,6 +4421,327 @@ Delete a single notification.
...
+                  "discount-url": "https://my-discount-url.com",
+                  "value": "50",
+                  "type": "percent",
+                  "is_active": "false",
+                  "tickets_number": "404",
+                  "min_quantity": "0",
+                  "max_quantity": "100",
+                  "valid_from": "2017-06-18T18:30:00+00:00",
+                  "valid_till": "2017-06-24T18:30:00+00:00",
+                  "used_for": "Event"
```
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @enigmaeth 
These also need to be dasherized.

File: docs/api/api_blueprint.apib [hunk 9]
```diff
@@ -4422,6 +4421,327 @@ Delete a single notification.
...
+              },
+              "attributes": {
+                  "tickets": null,
+                  "code": "DC101",
+                  "valid-from": "2017-06-18T18:30:00+00:00",
+                  "min-quantity": 0,
+                  "created-at": "2017-06-20T09:59:23.740772+00:00",
+                  "tickets-number": 404,
+                  "value": 50,
+                  "max-quantity": 100,
```
Comment (LuD1161) [inline docs/api/api_blueprint.apib]: @enigmaeth 
Here too , **tickets-number <= max-quantity**.

**Comment by niranjan94** [comment]
@LuD1161 why would tickets number be less than max.? Here tickets number is
the number of ticket for which this discount code can be used on.

**Comment by LuD1161** [comment]
@niranjan94 
So, if there are only 100 discount code tickets, so how can it be used for 404 tickets ?

**Comment by shubham-padia** [comment]
Max quantity is the no. Of max tickets per order, and tickets number is the max no. Of tickets per event.

**Comment by enigmaeth** [comment]
Okay, I was in the same notion of what niranjan said. I'll update the description of the tickets_number too.

**Comment by LuD1161** [comment]
@shubham-padia 
Why do we have 
>max no. Of tickets per event.

in the discount-code model ?

@enigmaeth 
What should it be updated to ?

**Comment by shubham-padia** [comment]
@LuD1161 @enigmaeth I think you misunderstood, Updated the comment, see again. ticket_numbers should not be less than max

**Pull request merged**